python/ql/src/Resources/FileNotAlwaysClosedQuery.qll

```diff
@@ -0,0 +1,137 @@
+/** Definitions for reasoning about whether files are closed. */
+
+import python
+import semmle.python.dataflow.new.internal.DataFlowDispatch
+import semmle.python.ApiGraphs
+
+/** A CFG node where a file is opened. */
+abstract class FileOpenSource extends DataFlow::CfgNode { }
+
+/** A call to the builtin `open` or `os.open`. */
+class FileOpenCall extends FileOpenSource {
+  FileOpenCall() {
+    this = [API::builtin("open").getACall(), API::moduleImport("os").getMember("open").getACall()]
+  }
+}
+
+private DataFlow::TypeTrackingNode fileOpenInstance(DataFlow::TypeTracker t) {
+  t.start() and
+  result instanceof FileOpenSource
+  or
+  exists(DataFlow::TypeTracker t2 | result = fileOpenInstance(t2).track(t2, t))
+}
+
+/**
+ * A call that returns an instance of an open file object.
+ * This includes calls to methods that transitively call `open` or similar.
+ */
+class FileOpen extends DataFlow::CallCfgNode {
+  FileOpen() { fileOpenInstance(DataFlow::TypeTracker::end()).flowsTo(this) }
+}
+
+/** A call that may wrap a file object in a wrapper class or `os.fdopen`. */
+class FileWrapperCall extends DataFlow::CallCfgNode {
+  DataFlow::Node wrapped;
+
+  FileWrapperCall() {
+    wrapped = this.getArg(_).getALocalSource() and
+    this.getFunction() = classTracker(_)
```

Comment:
I'm slightly puzzled by this very weak restriction. Do we really not require anything of the class that's wrapping it?

Reply:
It's what the original query did. 
The idea is that a class that takes a file in its constructor likely stores it in a field or similar. The impact of actually checking whether it does doesn't feel that high; there may be a few more results in some cases where the class isn't really a wrapper, but also potential for more FPs in some tricker cases where it is really a wrapper but in a way we miss.